File: dp/program_정수삼각형.py
from collections import deque

def game(triangle,dp):
    n = len(dp)
    queue = deque([[0,0]])
    dx = [0,1]
    dp[0][0] = triangle[0][0]
    while queue:
        x, y = queue.popleft()
        for i in range(2):
            nx = x + dx[i]
            ny = y + 1
            if  0<=ny<n and dp[ny][nx] < triangle[ny][nx] + dp[y][x]:
                dp[ny][nx] = triangle[ny][nx] + dp[y][x]
                queue.append([nx,ny])
    
    return max(dp[-1])

def solution(triangle):
    answer = 0
    cost = [[0]*len(x) for x in triangle]
    answer = game(triangle, cost)
    return answer




if __name__ == "__main__":

  solution([[7], [3, 8], [8, 1, 0], [2, 7, 4, 4], [4, 5, 2, 6, 5]])

File: dp/program_정수삼각형.java
import java.util.*;

class Solution {
    public int game(int[][] tri, int[][] dp){

        int n = tri.length;
        Queue<int []> queue = new LinkedList<>();
        queue.add(new int[] {0,0});
        int [] dx = {0,1};
        while(queue.size()>0){
            int[] cur = queue.poll();
            int x = cur[0];
            int y = cur[1];
            for(int i=0; i<2; i++){
                int nx = x + dx[i];
                int ny = y + 1;
                if (0<=ny && ny<n && dp[ny][nx] < tri[ny][nx] + dp[y][x]){
                    dp[ny][nx] = tri[ny][nx] + dp[y][x];
                    queue.add(new int[] {nx,ny});
                }
            }
        }

        return Arrays.stream(dp[n-1]).max().getAsInt();
    }

    public int solution(int[][] tri) {
        int[][] cost = new int [tri.length][tri.length];
        for(int i=0; i< tri.length; i++){
            Arrays.fill(cost[i],0);
        }
        cost[0][0] = tri[0][0];
        int answer = game(tri, cost);
        return answer;
    }
}